### DISCUSSION TITLE
Libraries and old time Mbed

### DISCUSSION THREAD

**Discussion by JohnK1987** [mbed-ce/mbed-os#239] [Ideas]
Hello,

I would like to do something with Mbed libraries but is necessary to create a space, so here are few options what i have in my mind.
1. Create a new organization only for libraries called MbedCE-Libraries. It will be separate from MbedCE but that allow let the MbedCE organization only for API examples and not so chaotic.
2. Use this organization (MbedCE) but when you use a repo for each library + example and potential examples for MbedOS APIs then there can be a chaos. Of course it can be solved by add readme to organization main page where we can put a list of libraries and MbedOS dependencies.
3.  something else?

Another point is current state of old libraries. For MbedCE it will be around add CmakeLists but harder will be covert the code of all old libraries to MbedOS6+ version without testing because i am not able to buy and test everything.

BTW i did an analyze and i found that most of changes what were done in MbedOS in the past are possible to revert back without break current state.
For example:
* old functions wait(float sec) and wait_(int ms) is possible to take back like a wrapper of thread_sleep_for
* old serial APIs like Serial, RawSerial, UARTSerial seem to be possible to use because SerialBase is still same like in MbedOS 5
* FunctionPointr was raplaced by Callback so there can be ez make a wrapper again
* Probably also TCPServer is still possible to use

Most of these changes break backward use of old libraries. So when we put these all back covered by a macro OLDTIMEBED (enabled via mbed_app.json) then we are able to use old libararies withou any code conversion at this time (can be done in future). Of course for price of old code in source.

What do you think? @multiplemonomials 
BR, Jan

**Comment by multiplemonomials** [comment]
I really like the idea of a separate organization for libraries.  I think we need a place for users to find & contribute to libraries that have been updated for Mbed CE!  Also that will help keep the Mbed CE github organization clear of clutter.  Maybe we should also make a 3rd organization for examples?  That way the mbed-ce org will be just the core 4-6 repos.

> So when we put these all back covered by a macro OLDTIMEBED (enabled via mbed_app.json)

Hmmm... I don't love that.  A lot of those functions were removed for a good reason, like wait() conflicting with OS function names and the old serial classes being literal cancer.  So, I would like it to be obvious that the user shouldn't use these for anything but adapting old libraries.  What if we made it a separate include file, like
```cpp
#include <mbed-legacy-shims.h>
```
that you had to include, and it would implement the old functions using modern Mbed?  I think that could work OK.  It could exist either as a separate library, or as an optional header in Mbed.

**Comment by JohnK1987** [comment]
The separate library does have the point because you have to edit library anyway and that not avoid unnecessary questions from low experienced users. But I agree with you we do not go this way.

For examples can be applied different techniques. For example multiple main_xxx.cpp files in one project and select final file of requested example via top lvl CmakeLists.txt
It is good for maintaining because you do not have so many repos also in case when you split examples to diffrent examples. Like one repo for examples around Drivers, one around USB and so on.

```
./MbedCE_Driver_Examples/
│
├── mbed-os/
├── SPI/
│	├── mabed_app.json
│	└── main.cpp
├── I2C/
│	├── mabed_app.json
│	└── main.cpp
├── PwmOut/
│	├── mabed_app.json
│	└── main.cpp
│
├── cmake-variants.yaml
└── CMakeLists.txt
```

So may i create new github organization MbedCE-libraries? Or you want do it by your self?
I suppose we can use MbedCE logo with additional label which will contain word 'libraries'.

**Reply by multiplemonomials** [reply]
You go ahead!  Just make me an admin on the org too.  

Also we should move this library from the mbed-ce org to the new org:
https://github.com/mbed-ce/RemoteIR
https://github.com/mbed-ce/RemoteIR-test-program

**Reply by multiplemonomials** [reply]
Here's the logo file if you need it (as png and Inkscape SVG): 
![mbed_ce](https://github.com/mbed-ce/mbed-os/assets/2099358/33c9f1f5-a045-4220-b945-55b8a1b203e7)

![mbed_ce](https://github.com/mbed-ce/mbed-os/assets/2099358/793b2b1c-33fe-406a-b01e-611c65ed77e4)

**Reply by JohnK1987** [reply]
Ok, I already created the org https://github.com/mbed-ce-libraries-examples. If you know about some necessary settings, please do it I do not know what is good to set there after start.
The name is universal so we can easily choose later about we place there just libraries with their examples or also examples for Mbed OS. 

My plan is import some libraries from old Mbed repos (with old commits if possible), add CmakeLists.txt and add readme for a description.

### Repo name:
* library - we hold old name of library inside the repo but the name of repo will look like `Mbed_lib_Lib-Name`
* example of a library - it could be `Mbed_lib_example_Lib-Name`
* example of something else - it can be `Mbed_example_Example-Name`

### Readme should contain info about:
* current state of the lib - already ported to latest Mbed or not and was tested with MbedCE
* general informations about the lib
* if the library will be simple then there can be also an example (in text). If the library will be more sophisticated then there will be a link to another repo with example or maybe only to different branch of same repo

### About Cmake in library:
I am still not so much experienced cmake user but i hope I understand some basics.
For example when we have a lib called Servo and contains Servo.h with Servo.cpp then CmakeLists.txt could look like
```
add_library(servo INTERFACE)

target_sources(servo
    INTERFACE
      Servo.h
      Servo.cpp
)

target_include_directories(servo
    INTERFACE
        .
)
```
Then just add to top lvl CmakeLists.txt `add_subdirectory(Servo)` directly after `add_subdirectory(mbed-os)`, right?

Please let me know what you like, what not and what is missing.

**Reply by JohnK1987** [reply]
First library was added to imagine of how it can look - https://github.com/mbed-ce-libraries-examples/DHT
I am not sure about licence. 
* Should i add the licence file into the repo when is a licence mentioned in header file?
* Should i add to the header file `Edit: Converted for MbedCE in 2024 by Johnyk...` ?

**Reply by multiplemonomials** [reply]
> Repo name

Hmm, why not just `<libname>` and `<libname>-Demo` or `<libname>-Example`?  I think that prefixing everything with Mbed_lib_ might get a bit annoying to look through (since they are already under the Mbed CE organization).

> Readme should contain info about

What you said sounds good to me!

> About Cmake in library

That's allllmost right.  You just need
```cmake
target_link_libraries(servo PUBLIC mbed-core-flags)
```
or if the library uses RTOS features:
```cmake
target_link_libraries(servo PUBLIC mbed-core-flags mbed-rtos-flags)
```
Otherwise the library won't be able to include mbed.h.  And be careful not to link the library to the `mbed-os` target; that will cause the library .a file to include all of Mbed and become huge.

**Reply by multiplemonomials** [reply]
Regarding the first library, that looks pretty good! We should definitely add the license that it says it's using. And I think adding a note that it was converted to Mbed CE (with a link to the original version) would be good.

**Reply by multiplemonomials** [reply]
Oh also!  Be careful with the CMake, the library should be a STATIC library, and you can include the sources in the add_library call.  So it should look like:
```cmake
add_library(dht11 STATIC
      DHT11.h
      DHT11.cpp
)
```
How you had it was the old way for non-CE Mbed, but using a static library is better because it means the library source will only be compiled once instead of once for every depender.

Also, be careful with the CMake file, it has to be CMakeLists.txt not CmakeLists.txt.

**Reply by JohnK1987** [reply]
Ok, I will edit it according to your recommendations.

> > About Cmake in library
> 
> That's allllmost right. You just need
> 
> ```cmake
> target_link_libraries(servo PUBLIC mbed-core-flags)
> ```
> 
> or if the library uses RTOS features:
> 
> ```cmake
> target_link_libraries(servo PUBLIC mbed-core-flags mbed-rtos-flags)
> ```
> 
> Otherwise the library won't be able to include mbed.h. And be careful not to link the library to the `mbed-os` target; that will cause the library .a file to include all of Mbed and become huge.

So this is wrong?
```
/****/
include(mbed-os/tools/cmake/app.cmake)

add_subdirectory(mbed-os)

add_subdirectory(DHT11)

project(MbedCE_example)

# add subdirectories and build targets here
add_executable(main main.cpp)

target_link_libraries(main mbed-os dht11) # Can also link to mbed-baremetal here
/* rest of script */
```

**Reply by multiplemonomials** [reply]
That file you posted looks fine, just don't forget mbed_set_post_build()!

**Reply by JohnK1987** [reply]
> ```cmake
> add_library(dht11 STATIC
>       DHT11.h
>       DHT11.cpp
> )
> ```
Oh, ok . So like this or?

```
add_library(dht11 STATIC
      DHT11.h
      DHT11.cpp
)

target_include_directories(dht11
    INTERFACE
        .
)
```

> Also, be careful with the CMake file, it has to be CMakeLists.txt not CmakeLists.txt.

Good point I didn't notice that

I already did some corrections (not all) and i will try create a template repo for add new library.

**Reply by multiplemonomials** [reply]
Just don't forget
```
target_link_libraries(dht11 PUBLIC mbed-core-flags)
```
as mentioned above!

**Reply by JojoS62** [reply]
I had problems when I started with cmake and tried to use static libs. When there are a handful of source files, it takes a second to compile, so I have no problem using them as interface lib. But I remember, this was an old discussion.

Anyway, I like the idea of creating repos that work with Mbed-CE.

**Reply by JohnK1987** [reply]
I saw the part of wiki page - [libraries-in-your-application]( https://github.com/mbed-ce/mbed-os/wiki/MbedOS-configuration#libraries-in-your-application) but it is not working like that without an include. What did i do wrong?

**My CMakeLists.txt inside the library:**
```
#[[https://github.com/mbed-ce/mbed-os/wiki/MbedOS-configuration#libraries-in-your-application]]

#[[ Set name of library in lowcase to the variable]]
set(INTERNAL_LIBRARY_NAME dht)

#[[ Here fill all .cpp files]]
add_library(${INTERNAL_LIBRARY_NAME} STATIC DHT.cpp)

#[[ Include .h files]]
target_include_directories(${INTERNAL_LIBRARY_NAME}
    INTERFACE
        .
)

#[[Link library to mbed-core-flags but if the library needs to use RTOS
 features, then comment out first option and uncoment the second option below]]
target_link_libraries(${INTERNAL_LIBRARY_NAME} PUBLIC mbed-core-flags)
#target_link_libraries(${INTERNAL_LIBRARY_NAME} PUBLIC mbed-core-flags mbed-rtos-flags)
```

**My top lvl CMakeLists.txt:**

```
cmake_minimum_required(VERSION 3.19)
cmake_policy(VERSION 3.19)

#### Initialize Mbed OS build system. ####

######################################################################################################
### Block of including .json files. Files of this block must be included before the app.cmake

#[[ Set path of mbed_app.json (necessary everytime if mbed_app.json is in project) ]]
set(MBED_APP_JSON_PATH mbed_app.json)

###--------------------------------------------------------------------------------------------------- 
#[[ This part is intended to custom targets and activates the custom_targets.json 
and upload method config otherwise functions below should be commented ]]

#[[ Here set path for custom_targets.json (not necessary if you choose another option) ]]
# set(CUSTOM_TARGETS_JSON_PATH custom_targets.json) 

#[[ Here set path for custom upload config .cmake (not necessary if you choose another option) ]]
# set(CUSTOM_UPLOAD_CFG_PATH ${CMAKE_SOURCE_DIR}/custom_targets/${MBED_TARGET}/${MBED_TARGET}.cmake)

### End of block
######################################################################################################

### include app.cmake (necessary everytime)
include(mbed-os/tools/cmake/app.cmake)


######################################################################################################
### Block of including project folders

#[[ In case of custom targers must be the folder included before mbed-os subdirectory
otherwise should be the next line commented]]
#add_subdirectory(custom_targets)

###--------------------------------------------------------------------------------------------------
## Add mbed-os subdirectory (necessary everytime)
add_subdirectory(mbed-os)

###--------------------------------------------------------------------------------------------------
## Add another subdirectory (if needed)
add_subdirectory(DHT)

### End of block
######################################################################################################

### Set up your project name (necessary everytime)
project(MbedCE_Example)

### add executable (necessary everytime)
add_executable(main main.cpp)

######################################################################################################
### Link libraries block

#[[link MbedOS and its libraries (necessary everytime)]]
target_link_libraries(main  mbed-os) 

### link user library (if needed)
target_link_libraries(main dht) 

#[[For more about this configuraion visit wiki page
https://github.com/mbed-ce/mbed-os/wiki/MbedOS-configuration#configuration-via-cmake-files]]

### End of block
######################################################################################################

### Set post build (necessary everytime)
mbed_set_post_build(main)

### Build finalize does necessary configuration for debug in most cases (necessary everytime)
mbed_finalize_build()
```

**Reply by multiplemonomials** [reply]
That looks right to me, what error are you getting?

**Reply by JohnK1987** [reply]
Ok, sorry my description was bad. With setup above it is working.
My point was about info on [libraries-in-your-application](https://github.com/mbed-ce/mbed-os/wiki/MbedOS-configuration#libraries-in-your-application)

I had to add the content below for avoiding `main.cpp:2:10: fatal error: DHT.h: No such file or directory`
```
#[[ Include .h files]]
target_include_directories(${INTERNAL_LIBRARY_NAME}
    INTERFACE
        .
)
```

Should I update the wiki about this line?

**Reply by multiplemonomials** [reply]
Oh yeah, that should probably be in there too.

**Reply by JojoS62** [reply]
great, the 
> target_link_libraries(${INTERNAL_LIBRARY_NAME} PUBLIC mbed-core-flags)
was missing in my attempts to use a static lib.

**Reply by JohnK1987** [reply]
I made 5 repos with already ported libraries - https://github.com/mbed-ce-libraries-examples In few of them I did another modification or correction.

This is still missing
> Also we should move this library from the mbed-ce org to the new org: https://github.com/mbed-ce/RemoteIR https://github.com/mbed-ce/RemoteIR-test-program

These libraries are also compatible with MbedOS6 and can be used in Keil Studio Cloud.
Maybe we can share this on forum.
